### PR TITLE
Minimum Containerized Application

### PR DESCRIPTION
**Pull request by mas393**
### Description
Changes allow the client and server to build and run in separate docker containers. From local testing, ground support can connect to a remote db instance.

To test, have the docker engine running and in the project root run the following in a terminal:
```
$ docker compose up --build
```
You should see output from the `react-client` and `express-server` containers. 

### TODO
- [ ] Test that the containerization works on Windows. (I've tested locally on Mac & Linux)
- [x] Add some documentation

### Notes
At the moment, the `Dockerfile` uses the `yarn install` command to avoid dependency resolution issues I was having with the main branch. We can switch back to using `npm install` once the issues are resolved.

Future work I did not attempt with this PR includes:
- A container for telemetry. If we are transitioning from python to rust, we should probably hold off on that for the time being.
- A container for a local mongo db. This is probably the next step, but we need to figure out a procedure for syncing local and cloud db instances first.
- Debugging tools. I'm unfamiliar with Typescript so I don't know what debugging tools and techniques are typically used. Feel free to add a comment about tools that should be integrated.

### CODE REVIEW

**Review by klemie — CHANGES_REQUESTED** [review]
Attempted docker build on windows got this: 

image

nodemon is not found. little strange, might be ts types error. Might need to add to devDependencies. Not sure why normal dependencies isn't enough.

In the future please create a branch rather than forking, cant pull your code testing the PR as easily. Also we only do branches in UVR

**Comment by klemie** [comment]
[Dockerization](https://trello.com/c/Qkl4tmXq/6-dockerization)

**Comment by mas393** [comment]
@klemie Thanks for trying the branch out. The issue was that on a fresh build, docker was overwriting the `node_modules` folder in the container when it created a bind mount of the `server` directory. I wasn't seeing the issue b/c I had a local `node_modules` folder which happened to contain the dependencies.

**Comment by klemie** [comment]
Tested update. Build is successful on windows. however the server container seems to think that express doesn't exist got a `MODULE_NOT_FOUND`.

Heres the logs:

```
2023-10-14 22:38:02 
2023-10-14 22:38:02 > service@1.0.0 start
2023-10-14 22:38:02 > npm run server
2023-10-14 22:38:02 
2023-10-14 22:38:03 
2023-10-14 22:38:03 > service@1.0.0 server
2023-10-14 22:38:03 > nodemon src/server.ts
2023-10-14 22:38:03 
2023-10-14 22:38:09 [nodemon] 3.0.1
2023-10-14 22:38:09 [nodemon] to restart at any time, enter `rs`
2023-10-14 22:38:09 [nodemon] watching path(s): *.*
2023-10-14 22:38:09 [nodemon] watching extensions: ts,json
2023-10-14 22:38:09 [nodemon] starting `ts-node src/server.ts`
2023-10-14 22:38:23 [nodemon] app crashed - waiting for file changes before starting...
2023-10-14 22:42:50 
2023-10-14 22:42:50 > service@1.0.0 start
2023-10-14 22:42:50 > npm run server
2023-10-14 22:42:50 
2023-10-14 22:42:50 
2023-10-14 22:42:50 > service@1.0.0 server
2023-10-14 22:42:50 > nodemon src/server.ts
2023-10-14 22:42:50 
2023-10-14 22:42:55 [nodemon] 3.0.1
2023-10-14 22:42:55 [nodemon] to restart at any time, enter `rs`
2023-10-14 22:42:55 [nodemon] watching path(s): *.*
2023-10-14 22:42:55 [nodemon] watching extensions: ts,json
2023-10-14 22:42:55 [nodemon] starting `ts-node src/server.ts`
2023-10-14 22:38:23 Error: Cannot find module 'express'
2023-10-14 22:38:23 Require stack:
2023-10-14 22:38:23 - /app/src/server.ts
2023-10-14 22:38:23     at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
2023-10-14 22:38:23     at Function.Module._resolveFilename.sharedData.moduleResolveFilenameHook.installedValue [as _resolveFilename] (/app/node_modules/@cspotcode/source-map-support/source-map-support.js:811:30)
2023-10-14 22:38:23     at Function.Module._load (node:internal/modules/cjs/loader:922:27)
2023-10-14 22:38:23     at Module.require (node:internal/modules/cjs/loader:1143:19)
2023-10-14 22:38:23     at require (node:internal/modules/cjs/helpers:119:18)
2023-10-14 22:38:23     at Object.<anonymous> (/app/src/server.ts:1:1)
2023-10-14 22:38:23     at Module._compile (node:internal/modules/cjs/loader:1256:14)
2023-10-14 22:38:23     at Module.m._compile (/app/node_modules/ts-node/src/index.ts:1618:23)
2023-10-14 22:38:23     at Module._extensions..js (node:internal/modules/cjs/loader:1310:10)
2023-10-14 22:38:23     at Object.require.extensions.<computed> [as .ts] (/app/node_modules/ts-node/src/index.ts:1621:12) {
2023-10-14 22:38:23   code: 'MODULE_NOT_FOUND',
2023-10-14 22:38:23   requireStack: [ '/app/src/server.ts' ]
2023-10-14 22:38:23 }
2023-10-14 22:43:02 Error: Cannot find module 'express'
2023-10-14 22:43:02 Require stack:
2023-10-14 22:43:02 - /app/src/server.ts
2023-10-14 22:43:02     at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
2023-10-14 22:43:02     at Function.Module._resolveFilename.sharedData.moduleResolveFilenameHook.installedValue [as _resolveFilename] (/app/node_modules/@cspotcode/source-map-support/source-map-support.js:811:30)
2023-10-14 22:43:02     at Function.Module._load (node:internal/modules/cjs/loader:922:27)
2023-10-14 22:43:02     at Module.require (node:internal/modules/cjs/loader:1143:19)
2023-10-14 22:43:02     at require (node:internal/modules/cjs/helpers:119:18)
2023-10-14 22:43:02     at Object.<anonymous> (/app/src/server.ts:1:1)
2023-10-14 22:43:02     at Module._compile (node:internal/modules/cjs/loader:1256:14)
2023-10-14 22:43:02     at Module.m._compile (/app/node_modules/ts-node/src/index.ts:1618:23)
2023-10-14 22:43:02     at Module._extensions..js (node:internal/modules/cjs/loader:1310:10)
2023-10-14 22:43:02     at Object.require.extensions.<computed> [as .ts] (/app/node_modules/ts-node/src/index.ts:1621:12) {
2023-10-14 22:43:02   code: 'MODULE_NOT_FOUND',
2023-10-14 22:43:02   requireStack: [ '/app/src/server.ts' ]
2023-10-14 22:43:02 }
2023-10-14 22:43:02 [nodemon] app crashed - waiting for file changes before starting...
```

also the front-end runs really slow so it should be a priority to switch to vite. Gonna put that in the queue for next sprint

File: compose.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+version: '3'
+
+services:
+  client:
+    container_name: react-client
+    build:
```
Comment (klemie) [inline compose.yaml:6]:
```suggestion
  client:
    container_name: react-client
    environment:
      - WDS_SOCKET_HOST=127.0.0.1 
      - CHOKIDAR_USEPOLLING=true
      - WATCHPACK_POLLING=true
    build:
```

File: compose.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    volumes:
+      - ./client/src:/app/src
+    ports:
+      - 3000:3000
+    depends_on:
+      - server
+
+  server:
+    container_name: express-server
+    build:
```
Comment (klemie) [inline compose.yaml:18]:
```suggestion
  server:
    container_name: express-server
    environment:
      - WDS_SOCKET_HOST=127.0.0.1 
      - CHOKIDAR_USEPOLLING=true
      - WATCHPACK_POLLING=true
    build:
```

**Review by klemie — CHANGES_REQUESTED** [review]
Found a solution for the hot reloading problem. It's kinda a catch all so we target all possible things holding up the binding. Works for me on windows, should double check with a fresh `docker compose up`

**Review by klemie — APPROVED** [review]
Hopefully CPU polling doesn't end the world 😬

**Pull request merged**